### PR TITLE
Update dev branch to use PyTorch 1.8 prerelease

### PR DESCRIPTION
**Pull request by fritzo**
Addresses #2754 
Blocking #2739, #2766

- @fehiepsi , @neerajprad could one of you review general changes?
- @stefanwebb could you take a look at the flows changes?

This pins our dev branch to track the latest PyTorch 1.8 nightly build, and fixes Pyro to be compatible with PyTorch 1.8.

This may be inconvenient for advanced users who use Pyro's dev branch, but it will help minimize breakage around the upcoming PyTorch 1.8 release by allowing us to release Pyro faster.

## Tested
- added `test_distributions.py::test_support_shape`
- added `test_distributions.py::test_is_discrete` and `is_not_discrete`
- xfailed some `collapse` tests; these will be fixed by #2739

### CODE REVIEW

File: pyro/distributions/transforms/cholesky.py [hunk 1]
```diff
@@ -87,8 +84,6 @@ class CholeskyTransform(Transform):
     positive definite matrix.
     """
     bijective = True
-    sign = +1
```
Comment (fritzo) [inline pyro/distributions/transforms/cholesky.py:90]: note `.sign` is only defined for scalar transforms, not multivariate transforms

File: tests/distributions/conftest.py [hunk 2]
```diff
@@ -262,20 +272,26 @@ def __init__(self, loc, scale):
...
+                    'test_data': [
+                        [[1.0, 0.0, 0.0],
+                         [-0.17332135, 0.98486533, 0.0],
+                         [0.43106407, -0.54767312, 0.71710384]],
+                        [[1.0, 0.0, 0.0],
+                         [-0.31391555, 0.94945091, 0.0],
+                         [-0.31391296, -0.29767500, 0.90158097]],
+                    ],
+                },
+            ]),
```
Comment (fritzo) [inline tests/distributions/conftest.py:294]: the old low-precision test data was failing the new `test_support_shape`, so I created new data

**Comment by fritzo** [comment]
@fehiepsi @neerajprad can we discuss this today? This PR is blocking development on pyro.distributions (#2739, #2766, ...)

File: tests/distributions/test_lkj.py [hunk 3]
```diff
@@ -18,7 +18,7 @@ def test_constraint(value_shape):
     value.diagonal(dim1=-2, dim2=-1).exp_()
     value = value / value.norm(2, dim=-1, keepdim=True)
 
-    assert (constraints.corr_cholesky_constraint.check(value) == 1).all()
+    assert (constraints.corr_cholesky.check(value) == 1).all()
```
Comment (neerajprad) [inline tests/distributions/test_lkj.py:21]: Likewise, this file can be removed now.

File: pyro/distributions/lkj.py [hunk 4]
```diff
@@ -4,13 +4,13 @@
...
 
 import torch
-from torch.distributions import constraints
 
-from pyro.distributions.constraints import corr_cholesky_constraint, corr_matrix
 from pyro.distributions.torch import Beta, TransformedDistribution
 from pyro.distributions.torch_distribution import TorchDistribution
 from pyro.distributions.transforms.cholesky import CorrMatrixCholeskyTransform, _vector_to_l_cholesky
 
+from . import constraints
```
[the remark this answers is not in the document]
Comment (fritzo) [reply pyro/distributions/lkj.py:12]: It looks like PyTorch has an `LKJCholesky` but neither an `LKJ` nor an `LKJCorrCholesky`. When I try to replace Pyro's `LKJCorrCholesky` with PyTorch's `LKJCholesk` the test_lkj.py tests fail. The code is quite different.

@fehiepsi do you understand the differences between these distributions, and can you suggest a deprecation plan?

File: tests/distributions/test_lkj.py (hunk 3 above)
Comment (fritzo) [reply tests/distributions/test_lkj.py:21]: Keeping file to test the `LKJ` distribution and backwards compatibility.

File: pyro/distributions/conjugate.py [hunk 5]
```diff
@@ -4,13 +4,15 @@
...
 from torch.distributions.utils import broadcast_all
 
-from pyro.distributions.torch import Beta, Binomial, Dirichlet, Gamma, Multinomial, Poisson
-from pyro.distributions.torch_distribution import TorchDistribution
 from pyro.ops.special import log_beta, log_binomial
 
+from . import constraints
+from .torch import Beta, Binomial, Dirichlet, Gamma, Multinomial, Poisson
+from .torch_distribution import TorchDistribution
+from .util import broadcast_shape
```
[the remark this answers is not in the document]
Comment (fritzo) [reply pyro/distributions/conjugate.py:14]: No; I have been slowly manually switching to relative imports.

File: pyro/distributions/lkj.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (neerajprad) [reply pyro/distributions/lkj.py:12]: Hmm..LkJCholesky and LKJCorrCholesky should be the same except that the sampling method should be more efficient. I think we can merge this for now and I'll take a look at the tests to see why that is failing.

**Review by neerajprad — APPROVED** [review]
LGTM. Thanks for making this change, @fritzo.

**Comment by fritzo** [comment]
@neerajprad thanks for reviewing! I'm happy merging this now and resolving LKJ differences in a follow-up PR. WDYT?

**Comment by neerajprad** [comment]
That sounds great. I'll take a look at the LKJ distribution differences after this PR merges.

**Comment by Balandat** [comment]
Hmm is this binary actually available (yet, or still)? I am getting the following, both in a fresh local conda installation on my mac, as well as in CI:
```
ERROR: Could not find a version that satisfies the requirement torch>=1.8.0.dev20210210 (from pyro-ppl)
ERROR: No matching distribution found for torch>=1.8.0.dev20210210
```

**Comment by Balandat** [comment]
Repro: 

```
conda create -n test_pyro
conda activate test_pyro
conda install pip
pip install git+https://github.com/pyro-ppl/pyro@dev
```

**Comment by fritzo** [comment]
@Balandat In linux you could try:
```
pip install --pre torch==1.8.0.dev20210210+cpu torchvision==0.9.0.dev20210210+cpu -f https://download.pytorch.org/whl/nightly/cpu/torch_nightly.html
```
And please let me know if you can suggest a better option for depending on PyTorch 1.8 prerelease.

**Pull request merged**